### PR TITLE
fix(runtime-host): reconnect after hosted target changes

### PR DESCRIPTION
**Pull request by hqhq1025**
## Summary

- detect when explicit hosted-target admission mutates the connection catalog or model inventory
- settle the invalidated owned Runtime Host and reconnect before `hosted.execution.start`
- compare provider default URLs canonically so unchanged targets do not restart

## Root cause

The first explicit DeepSeek target admission can commit model discovery. That invalidates the active backend and drains the owned Host. `runHostedExecution` then reused the now-closing connection for `hosted.execution.start`, producing `Runtime Host connection failed before execution settlement` before any model request.

## Verification

- `npm --workspace @maka/runtime-host test`: 841 passed, 0 failed
- focused hosted execution/target tests: 10 passed, 0 failed
- real-machine DeepSeek V4 Flash canary on `dhb`: settled `completed`, 24,708 tokens
- canary recorded `apply_patch` and `Read` as `outcome_committed`; proof file content matched exactly

The real-machine result uses the Terminal-Bench container and the same Maka eval shim as #2905. This PR contains no eval spec or machine-local configuration changes.

Generated with Codex assistance.

### CODE REVIEW

**Comment by hqhq1025** [comment]
Real-machine verification completed on the Terminal-Bench container with the #2905 Maka eval shim.

Before this fix, the same canary reproducibly returned `Runtime Host connection failed before execution settlement` immediately after first target admission. With `29638db6`, it settled `completed` with 24,708 tokens. The durable tool ledger records native `apply_patch` and `Read` as `outcome_committed`, the proof file matched exactly, and the model-facing tool inventory contained neither WebSearch nor WebFetch.

The repository Runtime Host suite also passed 841/841. This validates the reconnect boundary on a real owned Host, not only through mocks.

Generated with Codex assistance.

File: packages/runtime-host/src/client/hosted-execution-target.ts [hunk 1]
```diff
@@ -14,14 +15,15 @@ export async function configureHostedExecutionTarget(
...
-): Promise<void> {
+): Promise<boolean> {
   const before = await abortable(() => readRuntimeHostConnectionCatalog(connection), signal);
   const target = before.connections.find((candidate) => candidate.slug === input.connectionSlug);
   if (!target) throw new Error('Runtime Host connection is unavailable');
 
   const baseUrl = new URL(input.baseUrl).toString();
   const enabledModelIds = [...new Set([...target.enabledModelIds, input.model])];
-  const endpointChanged = target.baseUrl !== baseUrl;
+  const endpointChanged = new URL(effectiveBaseUrl(target)).toString() !== baseUrl;
```
Comment (Astro-Han) [inline packages/runtime-host/src/client/hosted-execution-target.ts]: [P2] Treat a missing current endpoint as a target change

`effectiveBaseUrl(target)` can legitimately be empty for an `openai-compatible` connection whose custom endpoint has not been stored yet. Calling `new URL('')` throws before this function can commit the supplied explicit `baseUrl`, so that valid connection can never be admitted through this path. Could the current endpoint be canonicalized safely, treating an empty or invalid existing value as “changed,” and could the final verification reuse the same comparator? A focused empty-current-URL regression should be enough. Please feel free to push back if explicit hosted targets intentionally exclude these connection records.

**Review by Astro-Han — APPROVED** [review]
Thanks for isolating this from the larger Eval work. The production change is small and sits at the right boundaries: target admission reports whether it committed a mutation, while `runHostedExecution` owns settling the invalidated Host generation and reconnecting before execution. The successful Runtime Host suite and real-machine canary also provide useful evidence for the main failure sequence.

I found one P2 edge case and left it inline: a valid connection without an effective base URL currently throws before the explicit target can write the supplied URL.

There is also a smaller cancellation-classification issue. If the user cancels after target admission but during the second Host connection, `connectOwnedRuntimeHost` can return `host_unresponsive`, and the new reconnect branch reports that as a startup failure rather than the existing `Hosted execution was cancelled` outcome. I would treat this as P3 because cancellation still succeeds and no authority or durable state is corrupted, but it would be worth applying the same abort check used by the initial connection path.

The tests could also be shorter and more precise:

- The mutation test counts close, settle, and reconnect separately, but does not prove the required `close → settle → reconnect → execute` ordering. A single shared event log would express the invariant more directly.
- The unchanged-target test proves that the helper returns `false`, but does not prove at the run level that execution stays on the first Host without reconnecting.
- The two reconnect tests repeat a substantial first-Host/catalog fixture. A small `mutatingFirstHost()` helper could remove most of that setup while preserving the success and reconnect-failure outcomes.
- A focused reconnect-cancellation case would cover the P3 behavior above.

I would keep the current production structure rather than add another abstraction. After fixing the inline edge case, tightening the cancellation result, and simplifying the tests, this should be ready to merge.

These are friendly suggestions, and I may be missing a constraint around empty connection endpoints or the cancellation contract, so please feel free to push back with that context.

AI-assisted review disclosure: Codex performed the read-only call-path analysis and independent review passes. Astro-Han reviewed the evidence, adjusted the severity to one P2 and one P3, and made the decision to approve and submit this review. The repository’s required independent human review remains a separate gate.

<details>
<summary>中文对照</summary>

感谢你把这个修复从较大的 Eval 工作中独立出来。生产代码规模很小，而且职责放置正确：target admission 负责报告是否提交了状态变化，`runHostedExecution` 负责让已经失效的 Host generation 完成退出，并在执行前重新连接。Runtime Host 全量测试和真实机器 canary 也为主要失败序列提供了有价值的证据。

我发现了一个 P2 边界情况，并通过 inline comment 标出：合法 connection 如果当前没有有效的 base URL，代码会在 explicit target 写入新 URL 之前提前抛错。

还有一个较小的取消分类问题。如果用户在 target admission 完成后、第二次连接 Host 的过程中取消，`connectOwnedRuntimeHost` 可能返回 `host_unresponsive`，而新增的重连分支会把它报告成启动失败，而不是现有的 `Hosted execution was cancelled`。我认为这是 P3，因为取消本身仍然生效，也没有破坏 authority 或持久状态；不过建议复用初始连接路径已有的 abort 检查。

测试也可以更短、更准确：

- mutation 测试分别统计了 close、settle 和 reconnect 次数，但没有证明必要的 `close → settle → reconnect → execute` 顺序。用一个共享事件序列会更直接地表达该不变量。
- unchanged-target 测试证明了 helper 返回 `false`，但没有在 run 层证明 execution 会继续使用第一个 Host 且不发生重连。
- 两个 reconnect 测试重复构造了较大的 first-Host/catalog fixture。可以提取一个小型 `mutatingFirstHost()` helper，在保留成功和重连失败结果的同时删除大部分重复 setup。
- 增加一个聚焦的 reconnect cancellation 测试即可覆盖上述 P3。

不建议为此增加新的抽象。修复 inline 边界、收紧取消结果并精简测试后，我认为就可以合并。

这些都是友好建议。如果 empty endpoint 或 cancellation contract 存在我没有看到的约束，欢迎直接说明或提出不同方案。

AI 辅助审查说明：Codex 进行了只读调用路径分析和多路独立审查；Astro-Han 复核了证据，将分级调整为一个 P2 和一个 P3，并决定批准和提交本审查。仓库要求的独立人工审查仍是单独的合并门槛。

</details>

**Comment by hqhq1025** [comment]
Addressed the follow-up review in amended commit `935225b2`:

- empty or invalid existing effective endpoints are now treated as changed, allowing the explicit target URL to be committed; final admission verification uses the same safe canonical comparator
- cancellation during the second owned-Host connection now preserves `Hosted execution was cancelled`
- tests now assert the exact `close → settle → reconnect → execute` ordering, prove unchanged targets execute on the first Host, share the mutating fixture, and cover reconnect cancellation

Verification after the amendment: focused tests 13/13; full Runtime Host suite 844/844; lint, format, typecheck, and `git diff --check` passed.

Generated with Codex assistance.
